lib/elixir/pages/anti-patterns/design-anti-patterns.md

```diff
@@ -411,6 +407,94 @@ end
 
 A common practice followed by the community is to make the non-raising version to return `{:ok, result}` or `{:error, Exception.t}`. For example, an HTTP client may return `{:ok, %HTTP.Response{}}` on success cases and a `{:error, %HTTP.Error{}}` for failures, where `HTTP.Error` is [implemented as an exception](`Kernel.defexception/1`). This makes it convenient for anyone to raise an exception by simply calling `Kernel.raise/1`.
 
+## Scattered side-effects
+
+#### Problem
+
+Elixir code tends to be explicit: it receives all inputs as arguments and returns all outputs. Furthermore, a given piece of code is clearer if, when called with the same inputs, it always returns the same output, without performing any other change in the system.
+
+For example, `String.split/2` fits this definition. It receives three arguments and, if the arguments do not change, the result is always the same:
+
+```elixir
+iex> String.split("hello world", " ")
+["hello", "world"]
+iex> String.split("hello world", " ")
+["hello", "world"]
+```
+
+On the other hand, `System.unique_integer/0` returns a different value, every time we call it:
+
+```elixir
+iex> System.unique_integer()
+-576460752303423478
+iex> System.unique_integer()
+-576460752303423446
+```
+
+We say that the `System.unique_integer/0` is *impure*, as it contains side-effects. When you invoke it, it changes the state of the program or the environment in a way that is not reflected on the output. On the other hand, we say `String.split/2` is *pure*. As a general rule, *pure* functions are easier to debug, test, and maintain then *impure* ones.
+
+Side-effects are essential when writing production software. Reading a file, logging a message, sending messages between processes, etc. are all side-effects. Writing code with side-effects is not an anti-pattern. However, you must avoid scattering side-effects throughout your functions.
+
+#### Example
```

Comment:
@whatyouhide @lucasvegi: I have added an example although I am not quite sure how happy I am with it. The discussion is complex and there is a nuance of exceptions being side-effects.

Reply:
@josevalim @whatyouhide I'm not sure if this example is showcasing 'scattering' well. If I'm not mistaken, in the example before refactoring, out of the 4 functions, two have side-effects (`recode!/1` and `encode_and_write!/2`). In the refactored code, the main change in terms of reducing 'scattering' was the elimination of the `encode_and_write!/2`, moving its body inside the `recode!/1` function, leaving it as the only one with side-effects among the remaining 3.

I understand that the second code undoubtedly has a better design than the first, I'm just not sure if it illustrates the concept of 'scattering' so clearly.

Reply:
Reading a file also has side-effects so all functions except the transformation one has side-effects. But I have a feeling the example could be better (or we discard the smell altogether).

Reply:
@josevalim, I thought a bit more and realized that I was confusing the granularity of what we consider 'scattering.' I was thinking more about scattering at a 'macro' level in the project, meaning something spread across many modules (e.g., CRUD operations). In your example, I understand that you have a finer-grained view, meaning scattered among functions within the same module.

So, with your observation that reading a file also has side-effects, I think the example is good, but it might even be better if we added comments to each function indicating whether it has side-effects or not. This way, it will be more evident to the reader (users of our catalog) that the scattering has decreased between the original and refactored code. Of course, this kind of comment wouldn't be suitable for production code, but I think it would work well for educational purposes.